Pane tabs: Scroll entire new tab into view

hrou0003 opened the pull request:
The state of the child bounds is not up-to-date when `scroll_to_item` gets triggered, causing the new tab to not scroll completely into view.

Closes #36317 

Release Notes:

- Fix an issue where a new tab is only partially visible on creation.

hrou0003 committed:
`scroll_to_item` was previously not taking into account the left side of the parent's bounds' causing it to not scroll the child completely into view

hrou0003 committed:
`scroll_to_item` previously did not take into account the left side of the parent's bounds, causing it to not scroll the child completely into view

hrou0003 committed:
`scroll_to_item` was previously not taking into account the left side of the parent's bounds causing it not scroll the child completely into view

hrou0003 committed:
Merge remote-tracking branch 'origin/hrou0003/fix-tab-creation-scrolling' into hrou0003/fix-tab-creation-scrolling

hrou0003 committed:
clippy + formatting

hrou0003 committed:
clippy + formatting

hrou0003 committed:
Merge remote-tracking branch 'origin/hrou0003/fix-tab-creation-scrolling' into hrou0003/fix-tab-creation-scrolling

MrSubidubi reviewed (changes requested):
Thanks for your shot at fixing this issue and appreciate you adding a test! I gave it a quick run and indeed, it does fix the issue for me.

However, a few things of note:
- If I cherry-pick the test onto the current main branch, it already passes. 
- Given what I can see above, we now only changed one of 4 similar looking patterns for handling the scroll offset related stuff, which seems off.
- A very similar issue still exists if I move focus out of the pane once and then re-enter that pane:

https://github.com/user-attachments/assets/3668ca7e-2b6a-4e01-982f-c33ead72eb7e

For a proper fix, I believe we should be fixing this case aswell. To me, it seems most reasonable that we either reserve space for the controls even if they are not shown or that we move the scroll position accordingly once the controls are shown again, so on focus in. What do you think?

MrSubidubi commented:
> Out of the 4 other patterns, I think the vertical-bottom case might be the only other one which needs addressing as using top/right already has the correct bounds - let me know if you think differently.

From an initial thought that sounds good. This would imply that we have a similar issue with vertical scrolling, so if you find a case where that fails aswell, that would be awesome.

hrou0003 committed:
Proposed changes to fix race condition between `notfiy`s update to `child_bounds` on `ScrollHandleState` and calling `scroll_to_item`

hrou0003 commented:
@MrSubidubi I've done a bit more debugging on this after trying to get a failing test, and the issue I'm now seeing is that the scroll state often gets updated after the call to `scroll_to_active_item` - meaning that it tries to scroll to what ends up being the second last tab. I've added some changes which seem to fix the issue, but I have a feeling there might be a better place to make to put the new call [here](https://github.com/zed-industries/zed/pull/36827/files#:~:text=if%20let%20Some%28scroll%5Fhandle%29%20%3D%20self%2Einteractivity%2Etracked%5Fscroll%5Fhandle%2Eas%5Fref%28%29%20%7B&text=scroll%5Fhandle%2Escroll%5Fto%5Factive%5Fitem%28%29%3B&text=%7D,-self%2Einteractivity%2Eprepaint), as it appears to cause minor issues switching between tabs.

Let me know if you have any thoughts or better ideas!

You can see see the behavior I've got and then the minor tab switching issue at the end of the recording

https://github.com/user-attachments/assets/d0f87b25-d856-4d56-95c6-d4bc4f192c36

hrou0003 committed:
Debugging for tab scroll

hrou0003 commented:
@MrSubidubi  Fair enough, I think I've got a pretty good grasp of the issue, but maybe not what's causing it. I think the assertions in the test I added in pane.rs aren't quite doing their job. However, when I add a line of debugging in scroll_to_item to show the state (against main): https://github.com/zed-industries/zed/pull/36827/files#diff-2c8b62d8a073267c39fea9d6135bb266cd28d447d7aa9accd78b16fec73d755eR3082

I get an output like this:
<img width="927" height="1010" alt="Screenshot 2025-09-08 at 21 04 18" src="https://github.com/user-attachments/assets/ba5de4ea-0de6-4de3-912d-fc7eebf41df3" />

You can also see that the tabs all have the same size, but the drop area is slightly bigger.

Based on this, I assumed that the `cx.notify()`'s in the pane occur after the call to `scroll_to_item`, and that's what my recent changes try to address. I have already tried deferring the call but that doesn't seem to fix the issue from my testing.

If you agree with this assessment, there's also a separate issue caused by making calls to scroll on focus in and out (to address your first comment), but I believe I have a fix for that as well, which I'll push.

MrSubidubi commented:
I assume this state is from the test? That definitely seems off, although admittedly, I haven't yet looked too deeply in the related code. 

Checking what's up with the test and the child items definitely seems like a good idea, other than that, I probably would look at how `scroll_offset()` and `max_offset()` behave in these scenarios on the associated `ScrollHandle` (after all, this is what directly influences the position). As mentioned, best case I'd like to be sure about the issue with the tab bar and then check how deep we have to reach within GPUI itself. Hope that seems reasonable.

hrou0003 committed:
- Removed `scroll_to_item` which can cause issues when scroll state is out of sync, replaced it with `scroll_to_active_item` which uses the scroll handle state's active item set by `update_active_item` to determine scrolling behavior and timing
- Track if the user has scrolled recently so we know if focus in and out should re-scroll to the active tab

hrou0003 committed:
Clippy

hrou0003 commented:
I've left in a failing test as it looks like the behavior of having the tab appear the buttons is intentional? It seems like we do want it appearing behind the buttons, but that doesn't seem quite right as it would contradict the original issue.

Any thoughts @MrSubidubi 

https://github.com/zed-industries/zed/actions/runs/17695521974/job/50294566737?pr=36827#step:6:772 https://github.com/zed-industries/zed/blob/d046016ef560b4829c0a4d5bd035cb85f9a66ba4/crates/collab/src/tests/integration_tests.rs#L6491

hrou0003 committed:
Remove assertion that was a bug + formatting

hrou0003 commented:
Thanks for taking the time to review, I think I did see that but didn't put it together that it could help - I'll try it out and see how it behaves.

I also noticed that uniform list seems to handle the vertical equivalent of the problem here: https://github.com/zed-industries/zed/blob/59fe6d2018a7064f0d0774178e2330c99c846d23/crates/gpui/src/elements/uniform_list.rs#L371C1-L393C22

Which is solves the problem in a similar way to mine but it might be worthwhile to align my implementation to that.

If it turns out that what I have now is the best option, I think we can avoid breaking the API if I instead rename `update_active_item` to `scroll_to_item` and then keep the modifications but rename the existing function, but do let me know if this won't work and I'll try your suggestion in the meantime.

hrou0003 committed:
Avoid breaking the `Div` API

hrou0003 commented:
I gave the anchor scroll a shot and realized by the time I had it 'working', that it it's impacted by a similar issue where `activate_item` is earlier in the cycle than render, which means we don't yet have an anchor to scroll to. I did also notice that there's the `Navigable` element and gave that a go but didn't have any success, I have a feeling it could still work but I don't think I can get it there - if you were keen to pair up on it I'd be happy to find some time.

Otherwise, I've pushed up some more changes to illustrate how we can avoid breaking the API with my existing changes.

MrSubidubi reviewed (changes requested):
Looks really good, thanks for bearing with me on this. Two minor things and we are good to merge!

Apologies for the slightly delayed review, was unfortunately out sick and thus unable to work.

hrou0003 committed:
Address comments

hrou0003 commented:
Hope you're feeling better! 

Thanks for you for your time and detailed feedback, I've pushed up those the changes and resolved the comments.

hrou0003 commented on crates/gpui/src/elements/div.rs:
```diff
@@ -3044,32 +3048,44 @@ impl ScrollHandle {
+        let mut state = self.0.borrow_mut();
 
-        let mut scroll_offset = state.offset.borrow_mut();
+        let active_item = match state.active_item {
```
Left the second match as there's a conflicting borrow between the offset and the state.

hrou0003 committed:
formatting

hrou0003 committed:
white space

MrSubidubi reviewed (approved):
I am finally feeling better, thank you. 

Looks good, at the risk of being repetitive, sorry for dragging this a bit and thank you very much for following all of this up and getting to the bottom of this! Much appreciated, thank you for your contribution!
